drivers/mhz19/mhz19_uart.c

```diff
@@ -0,0 +1,256 @@
+
+    /* Skip start byte and skip out of array bounds writes */
+    if ((dev->idx == 0 && byte == 0xff) || dev->idx >= MHZ19_BUF_SIZE) {
+        return;
+    }
+    /* Store byte and increment idx */
+    dev->rxmem[dev->idx++] = byte;
+    if (dev->idx == MHZ19_BUF_SIZE) {
+        mutex_unlock(&dev->sync);
+    }
+}
+
+int mhz19_init(mhz19_t *dev, const mhz19_params_t *params)
+{
+    dev->params = params;
+
+    DEBUG("mhz19: initializing device %p on UART %i\n",
+          (void *)dev, dev->params->uart);
+
+    mutex_init(&dev->mutex);
+    mutex_init(&dev->sync);
+    mutex_lock(&dev->sync);
+
+    dev->idx = 0;
+
+    /* Initialize UART interface */
+    if (uart_init(params->uart, MHZ19_UART_BAUDRATE, _mhz19_rx_cb, dev)) {
+        DEBUG("mhz19: UART initialization failed\n");
+        return MHZ19_ERR_INIT;
+    }
+
+    /* set auto calibration to a know value */
+    mhz19_set_auto_calibration(dev, params->auto_calibration);
+
+    DEBUG("mhz19: initialization complete\n");
+    return MHZ19_OK;
+}
+
+/*
+ * Do a raw send command to the sensor, without waiting for data.
+ *
+ * @param[in]  dev   The mhz19 device context
+ * @param[in]  out   the 9 bytes to transmit to the device
+ */
+static void mhz19_cmd(mhz19_t *dev, const uint8_t *in)
+{
+    /* First lock, guarantees no concurrent access to the UART device */
+    mutex_lock(&dev->mutex);
+
+    /* send the command */
+    uart_write(dev->params->uart, in, MHZ19_BUF_SIZE + 1);
+
+    /* Add some delay after executing command */
+    xtimer_usleep(MHZ19_TIMEOUT_CMD * US_PER_MS);
+
+    /* Unlock concurrency guard mutex */
+    mutex_unlock(&dev->mutex);
+}
+
+/*
+ * Do a raw send/receive exchange to the sensor. As exchanges between the
+ * MH-Z19 and the host always consists of 9 bytes in each direction, the size
+ * of the input and output arrays is fixed at 9 bytes here. The returned bytes
+ * from the MH-Z19 appear in mhz19_t::rxmem
+ *
+ * @param[in]  dev   The mhz19 device context
+ * @param[in]  out   the 9 bytes to transmit to the device
+ */
+static void mhz19_xmit(mhz19_t *dev, const uint8_t *in)
+{
+    /* timeout timer, for the case the sensor does not respond */
+    xtimer_t timer;
+
+    timer.callback = _mhz19_timeout;
+    timer.arg = dev;
+
+    /* Reset the buffer index to zero */
+    dev->idx = 0;
+
+    /* Send read command to the sensor */
+    uart_write(dev->params->uart, in, MHZ19_BUF_SIZE + 1);
+
+    /* Schedule timeout wakeup callback */
+    xtimer_set(&timer, MHZ19_TIMEOUT_CMD * US_PER_MS);
+
+    /* By locking the same mutex another time, this thread blocks until
+     * something else calls the unlock. In this case it has to wait for
+     * either the timeout xtimer callback or the UART ISR having received 9
+     * bytes. This way we have both a timeout scheduled and a regular wait
+     * for the UART to finish.
+     */
+    mutex_lock(&dev->sync);
+
+    /* lock mutex again for the next time */
+    mutex_lock(&dev->sync);
```

Comment:
Then you can drop this.

Reply:
Latest changes applies the suggestions and that works, but I had to add `mutex_unlock(..)` at the end to make it work.